Mostly-implement the create vpn call.

zenhack opened the pull request:
Tested using the mock backend for now. There are still a couple TODOs, but trying to keep prs to a managable size.

zenhack committed:
Add a type wrapper around the vpn names we'll pass to privop.

zenhack committed:
Add a check to make sure we're implementing the interfaces we mean to.

e.g. we didn't misspell a function name or something.

zenhack committed:
Use encoding/hex instead of fmt.Sscanf.

zenhack committed:
Start working on tracking state of vpns.

zenhack committed:
Move some validation logic into a common package.

The plan is to use this from vpnd as well.

zenhack committed:
Add a VpnStates argument to makeHandler.

also add a function to initialize a VpnStates.

zenhack committed:
Mock privops: validate the vpn name

zenhack committed:
Add some logic for allocating and freeing vpn ids/ports.

zenhack committed:
Write most of the create vpn api call.

zenhack commented:
Rebased this on top of #5, hopefully travis will pass now.

ianballou commented on cmd/hil-vpn-privop/main.go:
```diff
@@ -6,9 +6,10 @@ import (
 )
 
-// A regular expression matching legal vpn names.
 var vpnNameRegexp = regexp.MustCompile("^[-_a-zA-Z0-9]+$")
```
Do we still need this declaration here since you're importing the validate code?

zenhack replied:
Good catch, I meant to delete that...

ianballou commented on cmd/hil-vpnd/vpn-id.go:
```diff
@@ -0,0 +1,55 @@
+	return []byte(fmt.Sprintf("%d-%x", id.Port, id.Unique)), nil
+}
+
+func (id *VpnId) UnmarshalText(text []byte) error {
```
Do you mind giving a quick synopsis as to why we need to do the text marshalling and unmarshalling?  I think I could understand the code here better if I knew what the inputs and expected outputs of them will be.

zenhack replied:
We actually aren't using this at all, good catch -- I just pushed a commit that deletes them. I think I'd planned on using MarshalText here:

https://github.com/CCI-MOC/hil-vpn/pull/3/files#diff-f2f98c1532daa8f80c05622eb78cc4aaR57

...but ended up not doing so. The final format ended up being a bit different.

ianballou replied:
Oh ok, that clears up a lot of my confusion then

ianballou reviewed (commented):
This is looking fine to me, just a couple of questions.

zenhack committed:
Delete unused regexp.

This has been moved to another package.

zenhack committed:
Remove unused vpn-id

I think I'd planned on using these when generating the names we pass to
the PrivOps, but ended up not actually using it.

ianballou reviewed (approved):
Looks good to me now :+1:

naved001 commented on cmd/hil-vpnd/api.go:
```diff
@@ -35,6 +36,60 @@ func makeHandler(privops PrivOps) http.Handler {
+
+				// try to back out the change. TODO: think about the implications
+				// of what happens if we end up in this state; maybe we should
+				// blacklist the id for future use or something, just to be safe?
```
why would blacklisting the id help?

zenhack replied:
It wouldn't. At this point I only remember my reasoning well enough to know it didn't actually make any sense...

I removed the TODO from the comment.

zenhack replied:
As I context switch back in some of this is coming back to me. I believe I had been thinking *port*, rather than ID; the TODO was addressed in a later commit on #6:

https://github.com/CCI-MOC/hil-vpn/pull/6/commits/16a621cfdba95be160b6e0d54b92aa38b090d010

The issue is: suppose we fail to start openvpn, but it's been flagged to start on boot by systemd. Then, if we return the port to the free pool, we could end up creating another vpn for a different user using the same port, and possibly give them access to the wrong network.

naved001 replied:
okay, that explains it.

naved001 reviewed (commented):
looks good to me, I'll look at once again (sorry for the delay in getting to this)

naved001 commented on cmd/hil-vpnd/api_test.go:
```diff
@@ -1,14 +1,62 @@
+		t.Fatalf("Returned key disagrees with stored key; %v vs %v", results.Key, vpn.key)
+	}
+	if vpn.portNo != results.Port {
+		t.Fatalf("Returned poret disagrees with stored port; %v vs %v", results.Port,
```
typo: poret -> port

zenhack replied:
Fixed

naved001 commented on cmd/hil-vpnd/vpnstates.go:
```diff
@@ -0,0 +1,75 @@
+// Allocate a new port for a vpn.
+func (s *VpnStates) allocPort() (uint16, error) {
+	if len(s.FreePorts) == 0 {
+		return 0, ErrNoFreePorts
```
Just wanted to point out that port 0 is like a wild card port. Although we are sending the error along with it, but in case we forget to check the error when calling it, we may end up having a randomly allocated port. Maybe `return -1`, because that would definitely be an invalid port?

zenhack replied:
The type is uint16, so -1 as such isn't representable. I don't think there's a value we can return that isn't a valid port.

naved001 replied:
in that case let's just leave it as it is.

zenhack committed:
Fix a typo

zenhack committed:
Delete comment about possibly blacklisting ids.

Naved asked about this, looking back I think I just wasn't thinking
clearly at the time. We may leave the vpn in an inconsistent state, but
the id is (cryptographically) randomly generated and hasn't been handed
out yet, so it can't actually be used, and it isn't running, so I don't
think there's any real danger here.
